### DISCUSSION TITLE
Global events to log exceptions

### DISCUSSION THREAD

**Discussion by gulshan** [dotnet/runtime#91752] [Ideas]
The most common thing to do with the Exceptions is to log them. It can be called one of those "cross-cutting concerns". How about a few global events to do just that? There can be events for "All", "Caught" and "Uncaught" exceptions. These events can be fired on the catch-site or crash-site of the exceptions.

To be clear, I think the events should just have a read-only access of the exceptions. Meaning the events should not be able to "handle" or "consume" them, only log/record them. So, it would not prevent a crash or affect current exception handling mechanism in any way. But it will provide an additional centralized way to log the exceptions. Also, recording uncaught exceptions can be a fairly useful new feature.

I also think that these events should only be available to "executable" assemblies, not the libraries. Also, for the "caught" exceptions, the catch-site should be limited to the same assembly only, not the libraries.

Something similar can be found in the JS ecosystems.

**Comment by jnm2** [comment]
These exist:

- https://learn.microsoft.com/en-us/dotnet/api/system.appdomain.firstchanceexception (everything)
- https://learn.microsoft.com/en-us/dotnet/api/system.appdomain.unhandledexception (not caught)
- https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskscheduler.unobservedtaskexception (~thrown but not awaited)
- Caught by message loops:
  - https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.application.threadexception (Windows Forms)
  - https://learn.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher.unhandledexception (WPF)
  - https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.application.unhandledexception (UWP)
  - https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.application.unhandledexception (WinUI)
  
Sometimes exceptions are reported by other code that catches exceptions and continues:
- https://www.reactiveui.net/docs/handbook/default-exception-handler (ReactiveUI)
- https://learn.microsoft.com/en-us/aspnet/core/fundamentals/error-handling (ASP.NET)
- https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/handle-errors (Blazor)
